managed-ledger/src/main/java/org/apache/bookkeeper/mledger/impl/ManagedCursorImpl.java

```diff
@@ -553,6 +505,66 @@ protected void recoverFromLedger(final ManagedCursorInfo info, final VoidCallbac
         }
     }
 
+    private void recoverCursorByLedgerEntry(long entryId, VoidCallback callback, ManagedCursorInfo info,
+                                            LedgerHandle recoveredFromCursorLedger){
+        recoveredFromCursorLedger.asyncReadEntries(entryId, entryId, (rc1, lh1, seq, ctx1) -> {
+            if (log.isDebugEnabled()) {
+                log.debug("[{}} readComplete rc={} entryId={}", ledger.getName(), rc1, lh1.getLastAddConfirmed());
+            }
+            boolean tryPreviousEntry = Code.NoSuchEntryException == rc1 && entryId > 0;
+            if (tryPreviousEntry) {
+                recoverCursorByLedgerEntry(entryId - 1, callback, info, recoveredFromCursorLedger);
```

Comment:
If all the entries are not available, this might take a significant amount of time, since we're trying each and every entry. 

Even if it's not perfect, it would be better to try a binary-search approach to find the highest entry that is readable.

Reply:
I've changed the logic to binary search. And I have a question: will there be intermittent entry loss like this: `[x, null, x, null...]`. If this scenario exists, the binary search is not appropriate.